### PR TITLE
Database Configuration 

### PR DESCRIPTION
**Pull request by Fatmasiam**
Relates : #35

### CODE REVIEW

File: src/database/config/db_connection.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  port: params.port,
+  host: params.hostname,
+  database: params.pathname.split('/')[1],
+  ssl:params.hostname != 'localhost',
+  max: process.env.DB_MAX_CONNECTIONS || 2,
+  // return an error after 1 second if connection could not be established
+  connectionTimeoutMillis: 1000, 
+};
+
+module.exports = new Pool(options);
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js:27]: white Space

File: src/database/config/db_build.sql [hunk 2]
```diff
@@ -0,0 +1,44 @@
+BEGIN; 
+DROP TABLE IF EXISTS field, specialization, users, proposal, requirement;
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: you need to add ``` CASCADE ``` keyword

File: src/database/config/db_build.sql [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+CREATE TABLE field (
+	id  SERIAL PRIMARY KEY,
+	name varchar(50) NOT NULL
+);
+
+CREATE TABLE specialization (
+	id SERIAL PRIMARY KEY,
+	name varchar(30) NOT NULL,
+	fied_id INT REFERENCES field(id)
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: typo

File: src/database/config/db_build.sql [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+);
+
+CREATE TABLE users (
+	id SERIAL  PRIMARY KEY,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql:17]: why did you add ``` id ``` property while you can use ``` email ```  as primary key?

File: src/database/config/db_build.sql [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	firstname VARCHAR(20) NOT NULL,
+	lastname VARCHAR(20) NOT NULL,
+  email TEXT NOT NULL UNIQUE,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: ``` email ``` should be ``` varchar ``` not text 
#### As mention  here #6

File: src/database/config/db_build.sql [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  freelancer_url VARCHAR(50) NOT NULL,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: i think 50 chars not enough for something like ``` url ``` 
you can test any url and you can see

File: src/database/config/db_build.sql [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	photo_url VARCHAR(70) NOT NULL,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: the same thing as ``` freelancer_url ```

File: src/database/config/db_build.sql [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	birthday DATE NOT NULL,
+	mobile_number INT NOT NULL UNIQUE,
+	specalization_id INT REFERENCES specialization(id),
+  city VARCHAR(20),
+	password VARCHAR NOT NULL
+);
+
+CREATE TABLE proposal (
+	id SERIAL NOT NULL PRIMARY KEY ,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: ``` id ``` it's serial then it's  ``` not null ``` 
not null here is not necessary

File: src/database/config/db_build.sql [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	title varchar(50) NOT NULL,
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql:32]: you should update this issue #6 before making it in your P.R

File: src/database/config/db_build.sql [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	description varchar NOT NULL,
+	user_id INT REFERENCES users(id),
+	specalization_id int REFERENCES specialization(id),
+	contact_me TEXT NOT NULL
```
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: i think you should update that in your issue #6

File: src/database/config/db_build.sql (hunk 2 above)
Comment (kannanhassouna) [inline src/database/config/db_build.sql]: add white space here

File: src/database/config/db_build.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+const { readFileSync } = require('fs');
+const { join } = require('path');
+const connection = require('./db_connection');
+
+const reBuild = () => {
+  const filepath = join(__dirname,'db_build.sql');
```
Comment (kannanhassouna) [inline src/database/config/db_build.js:6]: i think you don't use ``` Eslint ``` , why ?

File: src/database/config/db_build.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  const sql = readFileSync(filepath).toString();
+  return connection.query(sql)
+};
+
+module.exports = reBuild;
```
Comment (kannanhassouna) [inline src/database/config/db_build.js]: white space

File: src/database/config/db_build.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
+const { readFileSync } = require('fs');
+const { join } = require('path');
+const connection = require('./db_connection');
+
+const reBuild = () => {
```
Comment (kannanhassouna) [inline src/database/config/db_build.js]: ``` reBuild ``` for what, you should be descriptive in naming like : ``` reBuildDB ```

File: src/database/config/db_connection.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+const params = parse(DB_URL);
+const [ user, password ] = params.auth.split(':');
+
+const options = {
+  user, 
+  password, 
+  port: params.port,
+  host: params.hostname,
+  database: params.pathname.split('/')[1],
+  ssl:params.hostname != 'localhost',
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js]: don't use loosely unequal

File: src/database/config/db_connection.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  max: process.env.DB_MAX_CONNECTIONS || 2,
+  // return an error after 1 second if connection could not be established
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js:24]: :+1: :+1: :+1: :+1: :1st_place_medal:

File: src/database/config/db_connection.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+const { parse } = require('url');
+const { Pool } = require('pg');
+require('dotenv').config();
+
+let DB_URL = '';
+if (process.env.NODE_ENV === 'test') {
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js]: white space here

File: src/database/config/db_connection.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+  DB_URL = process.env.TEST_DB_URL;
+}else if (process.env.NODE_ENV === 'production'){
+  DB_URL = process.env.DATABASE_URL;
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js]: you should use descriptive names like : 
  ``` PRODUCTION_DB_URL```,
  ``` DEVELOPMENT_DB_URL```,
  ``` TESTING_DB_URL```

File: src/database/config/db_connection.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+} else {
+  DB_URL = process.env.DB_URL;
+}
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js:11]: look, when i see alot of  ``` if else ``` then i use ``` switch ``` you can use if  you like it :  
```

switch (process.env.NODE_ENV) {
  case 'testing': DB_URL = process.env.TEST_DB_URL; break;
  case 'production': DB_URL = process.env.PRODUCTION_DB_URL; break;
  case 'development': DB_URL = process.env.DEVELOPMENT_DB_URL; break;
  default: throw new Error('Can not found the DB_URL !!!');
}
```

File: src/database/config/db_connection.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+const options = {
+  user, 
+  password, 
+  port: params.port,
+  host: params.hostname,
+  database: params.pathname.split('/')[1],
+  ssl:params.hostname != 'localhost',
+  max: process.env.DB_MAX_CONNECTIONS || 2,
+  // return an error after 1 second if connection could not be established
+  connectionTimeoutMillis: 1000,
```
Comment (kannanhassouna) [inline src/database/config/db_connection.js]: remove the comma here

**Review by kannanhassouna — CHANGES_REQUESTED** [review]

File: src/database/config/db_connection.js (hunk 1 above)
Comment (Fatmasiam) [reply src/database/config/db_connection.js:27]: done

File: src/database/config/db_build.sql (hunk 4 above)
Comment (Fatmasiam) [reply src/database/config/db_build.sql:17]: because of we will put id on payload

**Review by kannanhassouna — APPROVED** [review]

**Review by ahmedalmi95 — APPROVED** [review]

**Review by Angham116 — APPROVED** [review]

**Review by Anies12 — APPROVED** [review]
